### PR TITLE
[SPARK-27100][SQL] Use `Array` instead of `Seq` in `FilePartition` to prevent `StackOverflowError `

### PR DESCRIPTION
**Pull request by parthchandra**
## What changes were proposed in this pull request?

ShuffleMapTask's partition field is a FilePartition and FilePartition's 'files' field is a Stream$cons which is essentially a linked list. It is therefore serialized recursively.
If the number of files in each partition is, say, 10000 files, recursing into a linked list of length 10000 overflows the stack

The problem is only in Bucketed partitions. The corresponding implementation for non Bucketed partitions uses a StreamBuffer. The proposed change applies the same for Bucketed partitions.

## How was this patch tested?

Existing unit tests. Added new unit test. The unit test fails without the patch. Manual testing on dataset used to reproduce the problem.

### CODE REVIEW

**Comment by dbtsai** [comment]
add to whitelist

**Comment by dbtsai** [comment]
cc @dongjoon-hyun

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 1]
```diff
@@ -735,4 +744,60 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
         df1.groupBy("j").agg(max("k")))
     }
   }
+
+  //  a test with a single bucketed partition where the number of files in the partition is large
+  //  tests for the condition where the serialization of such a task may result in a stack overflow
+  //  if the files list is stored in a recursive data structure
+test("SPARK-27100 stack overflow: read bucketed data with large partitions") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: nit. indentation?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 2]
```diff
@@ -381,7 +381,7 @@ case class FileSourceScanExec(
       fsRelation: HadoopFsRelation): RDD[InternalRow] = {
     logInfo(s"Planning with ${bucketSpec.numBuckets} buckets")
     val filesGroupedToBuckets =
-      selectedPartitions.flatMap { p =>
+      selectedPartitions.toArray.flatMap { p =>
```
Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Does it make sense to apply the same change to `selectedPartitions.flatMap` in `createNonBucketedReadRDD(...)`?

Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Don't you want to put the `.toArray` much later here, like line 404?

**Review by srowen — CHANGES_REQUESTED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 2 above)
Comment (kiszk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Ah, although I thought `toArray.flatMap` was a kind of optimization. However, I cannot find `toArray.flatMap` in any other place.   
It looks good to keep the original code.

Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Yeah, I think just ensure the final result in the case class is an array. I'm not clear it would be faster to convert to an array and then maybe implicitly convert back to another type along the way, etc.

Comment (dbtsai) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: `@transient private lazy val selectedPartitions: Seq[PartitionDirectory]` is returning type of `Seq`. Since it's `lazy val` and used many times in the life cycle, why not we make it `@transient private lazy val selectedPartitions: Array[PartitionDirectory]`?

**Comment by dbtsai** [comment]
A question I would like to ask is when we should use `Array` instead of `Seq`. We extensively use `Seq` in our codebase because `Seq` is more generic interface. However, in this case, we don't have a control on what is being passed in, and the implementation can be a linked list resulting poor performance in many cases.

In stead of using `Seq` everywhere, should we use `Seq` in function input to keep it generic, but use `Array` in function return?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 2 above)
Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: That's probably fine if we need it to be an Array, or at least an IndexedSeq

Comment (parthchandra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: @dbtsai I like your suggestion but it requires that we either change the return type of `FileIndex.listFiles`  or convert the value returned by `relation.location.listFiles` to an array, which doesn't save us anything.  Changing the return type of `FileIndex.listFiles` is a bigger change that I would want to make at this point, given my relative inexperience with the code.

@srowen, 
```
 val filePartitions: Seq[FilePartition] = Seq.tabulate(bucketSpec.numBuckets) { bucketId =>
      FilePartition(bucketId, prunedFilesGroupedToBuckets.getOrElse(bucketId, Nil).toArray)
    }.toArray
```
doesn't compile 
```
/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:404: polymorphic expression cannot be instantiated to expected type;
[error]  found   : [B >: org.apache.spark.sql.execution.datasources.PartitionedFile]Array[B]
[error]  required: Seq[org.apache.spark.sql.execution.datasources.PartitionedFile]
[error]       FilePartition(bucketId, prunedFilesGroupedToBuckets.getOrElse(bucketId, Nil).toArray)
```
Scala isn't my first language so if you can help me with a way to satisfy the compiler, I'll be happy to make the change.

**Comment by srowen** [comment]
If one doesn't care, use Seq. Array is 'overkill', but simple, if the requirement is simply constant time random access. I think here that's what we want, more generally, as it won't be implemented as a LinkedList. `IndexedSeq` is OK if we can guarantee it, but, I think `Array` is fine and simple.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 2 above)
Comment (kiszk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: I also checked whether the return type can be changed to `Array[]`. I also realized that it is not a small change...

Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: `filePartitions` doesn't need to change. You declare it as a Seq but that's incompatible with the second `.toArray`. You just want the first `.toArray`. That works for me.

Comment (parthchandra) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Could `filePartitions` be a large list and run into the same serialization issue if we don't convert it into an array? Doesn't look likely, but I'd like to be sure. 
Also, @DBTsai suggested `  @transient private lazy val selectedPartitions: Array[PartitionDirectory] = {`. Do we want to do that?

**Comment by parthchandra** [comment]
Updated incorporating DBTsai's suggestion. Because this now affects the non bucketed read as well, added the unit test for the non bucketed read as well .

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 3]
```diff
@@ -735,4 +744,88 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+  }
+
+  //  a test with a single non-bucketed partition where the number of files in the partition is
+  //  large tests for the condition where the serialization of such a task may result in a stack
+  //  overflow if the files list is stored in a recursive data structure
+  test("SPARK-27100 stack overflow: read non bucketed data with large partitions") {
+    testLargeFilePartitionStackOverflow(true)
+  }
+
+  private def testLargeFilePartitionStackOverflow ( isBucketed : Boolean) {
```
Comment (srowen) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: Nit: remove spaces around the args here

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 4]
```diff
@@ -735,4 +744,88 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+  //  if the files list is stored in a recursive data structure
+  test("SPARK-27100 stack overflow: read bucketed data with large partitions") {
+    testLargeFilePartitionStackOverflow(true)
+  }
+
+  //  a test with a single non-bucketed partition where the number of files in the partition is
+  //  large tests for the condition where the serialization of such a task may result in a stack
+  //  overflow if the files list is stored in a recursive data structure
+  test("SPARK-27100 stack overflow: read non bucketed data with large partitions") {
+    testLargeFilePartitionStackOverflow(true)
```
Comment (srowen) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: Should one of these be false? how about just one test that calls this for true, false? You don't need three methods then, just one.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 5]
```diff
@@ -735,4 +744,88 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+    } catch {
+      case ex: StackOverflowError =>
+        val bucketingType = if (isBucketed) {
+          "bucketed"
+        } else {
+          "non-bucketed"
+        }
+        fail("Stack Overflow Exception in serializing task to read partitioned %s tables"
+          .format(bucketingType))
+      case _ => fail("Exception in serializing task to read partitioned tables")
```
Comment (srowen) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: This would mask the exception. Just let the exception go.

**Review by srowen — CHANGES_REQUESTED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala (hunk 3 above)
Comment (parthchandra) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala (hunk 4 above)
Comment (parthchandra) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: Yes you're right. Careless mistake on my part. And as it turns out the unit test was broken for the non-bucketed case. Fixed and unified the test cases into one.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala (hunk 5 above)
Comment (parthchandra) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 6]
```diff
@@ -46,7 +55,8 @@ class BucketedReadWithoutHiveSupportSuite extends BucketedReadSuite with SharedS
 }
 
 
-abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
+abstract class BucketedReadSuite extends QueryTest with SQLTestUtils with PrivateMethodTester {
+
```
Comment (dbtsai) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: remove this new line.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 7]
```diff
@@ -735,4 +745,100 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+      val inputRDD = if (isBucketed) {
+        // Create a Bucketed RDD. This is a private method so we need to call this indirectly.
+        val createBucketedReadRDD = PrivateMethod[RDD[InternalRow]]('createBucketedReadRDD)
+
+        fileSource invokePrivate createBucketedReadRDD(bucketSpec,
+          (file: PartitionedFile) => Seq(InternalRow(1)).toIterator,
+          Array(partitionDirectory),
+          fakeHadoopFsRelation)
+      } else {
+        // Create a Bucketed RDD. This is a private method so we need to call this indirectly.
```
Comment (dbtsai) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: Is this comment correct?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 8]
```diff
@@ -735,4 +745,100 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+      val partitionValues = InternalRow.apply(Array("a"))
+
+      val schema = new StructType()
+      val fakeHadoopFsRelation = new HadoopFsRelation(null, schema, schema, null,
+        new ParquetFileFormat, null)(sparkSession)
+      val optionalBucketSet = null
+      val bucketSpec = new BucketSpec(1, Seq("a"), Seq("b"))
+      val files = (0 to numFilesInPartition - 1).toStream.map { i =>
+        new FileStatus(10, false, 1, 512, 1000,
+          new Path(s"file${i}_0.zzz"))
```
Comment (dbtsai) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: move this line up.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 9]
```diff
@@ -735,4 +745,100 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+      }
+      val partitionDirectory = PartitionDirectory(partitionValues, files);
+
+      val fileSource =
+        FileSourceScanExec(fakeHadoopFsRelation,
```
Comment (dbtsai) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: move this line up.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 10]
```diff
@@ -735,4 +745,100 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+
+        fileSource invokePrivate createBucketedReadRDD(bucketSpec,
+          (file: PartitionedFile) => Seq(InternalRow(1)).toIterator,
+          Array(partitionDirectory),
+          fakeHadoopFsRelation)
+      } else {
+        // Create a Bucketed RDD. This is a private method so we need to call this indirectly.
+        val createNonBucketedReadRDD = PrivateMethod[RDD[InternalRow]]('createNonBucketedReadRDD)
+
+        fileSource invokePrivate createNonBucketedReadRDD(
```
Comment (dbtsai) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: We typically don't use this style. `fileSource.invokePrivate.createNonBucketedReadRDD`

**Review by dbtsai — COMMENTED** [review]
LGTM, just minor stylings.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 11]
```diff
@@ -735,4 +745,100 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+    if (spark.sparkContext.conf.get(CATALOG_IMPLEMENTATION) == "hive") {
+      assert(unbucketedCount == numFilesInPartition)
+    } else {
+      // default parallelism of 2 causes the files to be split into two partitions
+      assert(unbucketedCount == numFilesInPartition / 2)
+    }
+    createAndSerializeTask(unbucketedInputRDD, "non-bucketed")
+
+  }
+
```
Comment (dbtsai) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: remove extra new line

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 12]
```diff
@@ -735,4 +745,100 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+    val unbucketedInputRDD = getInputRDD(newSession, false);
+    val unbucketedCount = unbucketedInputRDD.partitions(0).asInstanceOf[FilePartition].files.length;
+    if (spark.sparkContext.conf.get(CATALOG_IMPLEMENTATION) == "hive") {
+      assert(unbucketedCount == numFilesInPartition)
+    } else {
+      // default parallelism of 2 causes the files to be split into two partitions
+      assert(unbucketedCount == numFilesInPartition / 2)
+    }
+    createAndSerializeTask(unbucketedInputRDD, "non-bucketed")
+
```
Comment (dbtsai) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: remove extra line

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 13]
```diff
@@ -735,4 +745,100 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+        fileSource invokePrivate createNonBucketedReadRDD(
+          (file: PartitionedFile) => Seq(InternalRow(1)).toIterator,
+          Array(partitionDirectory),
+          fakeHadoopFsRelation)
+
+      }
+      inputRDD
+    }
+
+    def createAndSerializeTask(inputRDD: RDD[InternalRow], bucketingType: String) {
```
Comment (dbtsai) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: `def tryCreateAndSerializeTask(inputRDD: RDD[InternalRow], bucketingType: String) = {`

**Review by dbtsai — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 14]
```diff
@@ -735,4 +746,97 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+          null,
+          Option(optionalBucketSet),
+          Seq.empty,
+          Option(new TableIdentifier("stackOverflow")))
+
+      val inputRDD = if (isBucketed) {
+        // Create a Bucketed RDD. This is a private method so we need to call this indirectly.
+        val createBucketedReadRDD = PrivateMethod[RDD[InternalRow]]('createBucketedReadRDD)
+
+        fileSource.invokePrivate(createBucketedReadRDD(bucketSpec,
```
Comment (kiszk) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: nit: Is it better to move [this line](https://github.com/apache/spark/pull/24865/files#diff-c410498f1467cb430723b2eca2536795R764) to here?

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @parthchandra .
According to @dbtsai 's comment, I made another style-fix PR to you, https://github.com/parthchandra/spark/pull/1 . Could you review and merge that first?

**Comment by dbtsai** [comment]
Any more feedback? If this looks good to all of you, I'll merge it when the build finishes. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala (hunk 14 above)
Comment (parthchandra) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: Assuming you mean the line `val bucketSpec = BucketSpec(1, Seq("a"), Seq("b"))`. Done

**Comment by parthchandra** [comment]
I made the last change based on @kiszk's suggestion. @dongjoon-hyun the style fix you provided is easily the best way of providing feedback I have ever gotten in a code review :)

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 15]
```diff
@@ -735,4 +745,92 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+        HadoopFsRelation(null, schema, schema, null, new ParquetFileFormat, null)(sparkSession)
+      val optionalBucketSet = null
+      val files = (0 until numFilesInPartition).toStream.map { i =>
+        new FileStatus(10, false, 1, 512, 1000, new Path(s"file${i}_0.zzz"))
+      }
+      val partitionDirectory = PartitionDirectory(partitionValues, files)
+
+      val fileSource = FileSourceScanExec(fakeHadoopFsRelation,
+          null,
+          schema,
```
Comment (dbtsai) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: this should be two spaces indentation.

**Comment by dongjoon-hyun** [comment]
@dbtsai . Please update the PR title~

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 16]
```diff
@@ -735,4 +745,92 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+    // attempt to serialize it in the task.
+    val bucketedCount = bucketedInputRDD.partitions(0).asInstanceOf[FilePartition].files.length
+    assert(bucketedCount == numFilesInPartition)
+    tryCreateAndSerializeTask(bucketedInputRDD, "bucketed")
+
+    // Non Bucketed partitions
+    // create a new session so we can change the config and force a large number of files
+    // in each partition
+    val newSession = spark.cloneSession()
+    newSession.conf.set("spark.sql.files.openCostInBytes", 0)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: Shall we use `withSQLConf("spark.sql.files.openCostInBytes" -> "0") { ... }` instead of creating a new session? `spark.sql.files.openCostInBytes` is a dynamic configuration, so we can change it in a session. Also, `withSQLConf` will recover it back after this test.

Comment (dbtsai) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: +1

Comment (parthchandra) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: Nice. I struggled for a bit to find a way to change conf temporarily for a single test. This fits the bill nicely. Change made

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala (hunk 15 above)
Comment (parthchandra) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: done

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 17]
```diff
@@ -735,4 +745,93 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+    // check to make sure we've created the a big enough file partition.
+    // also guarantees that the 'files' Stream is initialized before we
+    // attempt to serialize it in the task.
+    val bucketedCount = bucketedInputRDD.partitions(0).asInstanceOf[FilePartition].files.length
+    assert(bucketedCount == numFilesInPartition)
+    tryCreateAndSerializeTask(bucketedInputRDD, "bucketed")
+
+    // Non Bucketed partitions
+    // create a new session so we can change the config and force a large number of files
+    // in each partition
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: Thank you for updating, could you remove line 822 ~ 823, too?

**Review by kiszk — APPROVED** [review]
except @dongjoon-hyun 's last comment

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 18]
```diff
@@ -735,4 +745,93 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+        Option(new TableIdentifier("stackOverflow")))
+
+      val inputRDD = if (isBucketed) {
+        // Create a Bucketed RDD. This is a private method so we need to call this indirectly.
+        val createBucketedReadRDD = PrivateMethod[RDD[InternalRow]]('createBucketedReadRDD)
+        val bucketSpec = BucketSpec(1, Seq("a"), Seq("b"))
+
+        fileSource.invokePrivate(createBucketedReadRDD(bucketSpec,
+          (_: PartitionedFile) => Seq(InternalRow(1)).toIterator,
+          Array(partitionDirectory),
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: Ur, unfortunately, this test case becomes unable to detect the regression after changing from `Stream(partitionDirectory)` to `Array(partitionDirectory)`.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 19]
```diff
@@ -735,4 +745,93 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+          fakeHadoopFsRelation))
+      } else {
+        // Create a non Bucketed RDD. This is a private method so we need to call this indirectly.
+        val createNonBucketedReadRDD = PrivateMethod[RDD[InternalRow]]('createNonBucketedReadRDD)
+
+        fileSource.invokePrivate(createNonBucketedReadRDD(
+          (_: PartitionedFile) => Seq(InternalRow(1)).toIterator,
+          Array(partitionDirectory),
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: Here, too.

**Comment by dongjoon-hyun** [comment]
Without the patch, this test case fails with `IllegalArgumentException` due to `Array(partitionDirectory)` type mismatch. (Of course, it's natural.)
```
[info] - SPARK-27100 stack overflow: read data with large partitions *** FAILED *** (5 milliseconds)
[info]   java.lang.IllegalArgumentException: Can't find a private method named: createBucketedReadRDD
```

If we use `Array(partitionDirectory).toSeq` or `Seq(partitionDirectory)`, this test case cannot detect the regression. In other words, it succeeds without the patch. 

The test case fails only when we use the original form `Stream(partitionDirectory)`. However, `Stream(partitionDirectory)` will raise `IllegalArgumentException` with the patch due to the same type mismatch reason.

To sum up, it seems that we cannot have a meaningful test case at this level. To prevent a future regression, we need a higher level test case. Could you try to add that, @parthchandra ?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala (hunk 18 above)
Comment (parthchandra) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: Yes. Essentially the fix is in the signature change for the `create*BucketedReadRDD` functions. So the only way to invoke them is by passing in `Array(partitionDirectory)`. I see no way for this unit test to demonstrate the original bug. Now, if someone change things back to Seq, this unit test will still pass. 

The following will demonstrate the bug :
```
  test("read bucketed data with large partitions") {
    val nCount = 12000
    // reshuffle data so that many small files are created
    val nShufflePartitions = 6000
    // and with one table partition, should result in 6000 files in one partition
    val nPartitions = 1
    val nBuckets = 2
    val dfPartitioned = (0 until nCount)
      .map(i => (i % nPartitions, i % nBuckets, i.toString)).toDF("i", "j", "k")

    try {
      withTable("bucketed_table") {
        dfPartitioned.repartition(nShufflePartitions)
          .write
          .format("parquet")
          .partitionBy("i")
          .bucketBy(nBuckets, "j")
          .saveAsTable("bucketed_table")

        val table = spark.table("bucketed_table")
        val nValues = table.select("j", "k").count()
        assert(nValues == nCount)
      }
    } catch {
      case e : Exception => fail("Failed due to exception: " + e)
    }
  }
```
as well as the fix but takes a really longtime to run since it creates 6K temporary files.

Our other choice is to revert back to the original fix and apply @srowen's suggestion. 
It simplifies the unit test considerably and in fact we can remove the nonBucketed case since it is no longer affected by the fix (it never had a problem to start with).

Thoughts?

**Comment by parthchandra** [comment]
Yes, I tried the same experiments too. Take a look at the test I pasted in my previous response. It takes several minutes to run which is not desirable.

**Comment by parthchandra** [comment]
OK. Replaced the unit test with a higher level unit test. I added a query to test for both bucketed and non-bucketed tables which doubles the test time to about 5 min.
So the test is ignored for now.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala [hunk 20]
```diff
@@ -735,4 +735,54 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
...
+
+  //  A test with a partition where the number of files in the partition is
+  //  large. tests for the condition where the serialization of such a task may result in a stack
+  //  overflow if the files list is stored in a recursive data structure
+  //  This test is ignored because it takes long to run (~3 min)
+  ignore("SPARK-27100 stack overflow: read data with large partitions") {
+    val nCount = 12000
+    // reshuffle data so that many small files are created
+    val nShufflePartitions = 6000
+    // and with one table partition, should result in 6000 files in one partition
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: Could you increase this like the following? The current value, `6000`, fails within IDE (IntelliJ) environment, but succeed SBT environment (`build/sbt "sql/testOnly *.BucketedReadWithoutHiveSupportSuite").
```
+    val nShufflePartitions = 10000
+    // and with one table partition, should result in 10000 files in one partition
```

Comment (parthchandra) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: Done. So many things to remember to check :) Thank you for being so careful.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: Ur, @parthchandra . I gave you two lines including the comment.
Could you fix the comment, too?

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala]: BTW, I didn't ask the `nCount`. Do you need to upgrade that for some reason?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

@dbtsai . Could you review this again and do final sign-off, please?

**Comment by dbtsai** [comment]
LGTM too. Merged into master. 

@parthchandra can you raise the same changes to branch 2.4 (this PR can not be cleanly merged into branch 2.4)?

Thanks all.

**Comment by parthchandra** [comment]
Thanks @dbtsai, @dongjoon-hyun, @srowen, @kiszk for the excellent review comments.
